### Sport Analysis - over 1.5 Frequemcy

we are analysing over all the availble soccer features on 28th of Sept, 2024.

Let's find the likely of 0.5 at the end of the match

Api Provider => Api Football (www.api-football.com) 

### First lets get our sports data

In [79]:
import os, time, json, requests
from dotenv import load_dotenv
import pandas as pd

load_dotenv() #load environment variables in the os

True

In [73]:
date = "2024-09-28"
url = f"https://v3.football.api-sports.io/fixtures?date={date}"
json_file = f'football_{date}.json'

In [81]:
api_key = os.getenv("football-api-key")

In [51]:
# # Get Fixtures by date then save fixtures
# headers = {
#     'x-rapidapi-host': "v3.football.api-sports.io",
#     'x-rapidapi-key': "api_key"
# }


# response = requests.get(url, headers=headers)    
# if response.status_code == 200:
#     # Parse the response as JSON
#     fixtures = response.json()

#     with open(json_file, 'w') as file:
#         json.dump(fixtures, file, indent=4)
    
#     print("Data fetched successfully!")
    
# else:
#     print(f"Failed to fetch data. Status code: {response.status_code}")


Data fetched successfully!


**We had our data file stored so we can always work with it next time without hitting the api**

In [52]:
#read json file

with open(json_file) as file:
        data = json.load(file)

In [53]:
#get fixtures as data response

fixtures = data['response']
fixtures

[{'fixture': {'id': 1158871,
   'referee': 'J. Baliño',
   'timezone': 'UTC',
   'date': '2024-09-28T00:00:00+00:00',
   'timestamp': 1727481600,
   'periods': {'first': 1727481600, 'second': 1727485200},
   'venue': {'id': 83,
    'name': 'Estadio Juan Bautista Gargantini',
    'city': 'Mendoza, Provincia de Mendoza'},
   'status': {'long': 'Match Finished', 'short': 'FT', 'elapsed': 90}},
  'league': {'id': 128,
   'name': 'Liga Profesional Argentina',
   'country': 'Argentina',
   'logo': 'https://media.api-sports.io/football/leagues/128.png',
   'flag': 'https://media.api-sports.io/flags/ar.svg',
   'season': 2024,
   'round': '2nd Phase - 16'},
  'teams': {'home': {'id': 473,
    'name': 'Independ. Rivadavia',
    'logo': 'https://media.api-sports.io/football/teams/473.png',
    'winner': None},
   'away': {'id': 478,
    'name': 'Instituto Cordoba',
    'logo': 'https://media.api-sports.io/football/teams/478.png',
    'winner': None}},
  'goals': {'home': 1, 'away': 1},
  'score'

In [10]:
#let's view the structure of a single feature

fixtures[0]

{'fixture': {'id': 1158871,
  'referee': 'Jorge Balino, Argentina',
  'timezone': 'UTC',
  'date': '2024-09-28T00:00:00+00:00',
  'timestamp': 1727481600,
  'periods': {'first': 1727481600, 'second': 1727485200},
  'venue': {'id': 83,
   'name': 'Estadio Juan Bautista Gargantini',
   'city': 'Mendoza, Provincia de Mendoza'},
  'status': {'long': 'Match Finished', 'short': 'FT', 'elapsed': 90}},
 'league': {'id': 128,
  'name': 'Liga Profesional Argentina',
  'country': 'Argentina',
  'logo': 'https://media.api-sports.io/football/leagues/128.png',
  'flag': 'https://media.api-sports.io/flags/ar.svg',
  'season': 2024,
  'round': '2nd Phase - 16'},
 'teams': {'home': {'id': 473,
   'name': 'Independ. Rivadavia',
   'logo': 'https://media.api-sports.io/football/teams/473.png',
   'winner': None},
  'away': {'id': 478,
   'name': 'Instituto Cordoba',
   'logo': 'https://media.api-sports.io/football/teams/478.png',
   'winner': None}},
 'goals': {'home': 1, 'away': 1},
 'score': {'halftime'

### Now let's process the json response to a Pandas DataFrame for further Processing and Analysis 

In [83]:
fixturesDF = pd.json_normalize(fixtures, sep='.')
fixturesDF

fixture.id             fixture.referee fixture.timezone  \
0        1158871                   J. Baliño              UTC   
1        1174518                    J. Matto              UTC   
2        1206084  Oscar Mejia Garcia, Mexico              UTC   
3        1223155                  J. Benítez              UTC   
4        1278249                        None              UTC   
...          ...                         ...              ...   
1739     1168633                 I. Petrunok              UTC   
1740     1223151                        None              UTC   
1741     1285218                        None              UTC   
1742     1306667                        None              UTC   
1743     1234120                        None              UTC   

                   fixture.date  fixture.timestamp  fixture.periods.first  \
0     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
1     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
2     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
3     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
4     2024-09-28T00:00:00+00:00         1727481600                    NaN   
...                         ...                ...                    ...   
1739  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1740  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1741  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1742  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1743  2024-09-28T23:45:00+00:00         1727567100           1.727567e+09   

      fixture.periods.second  fixture.venue.id  \
0               1.727485e+09              83.0   
1               1.727485e+09           20581.0   
2               1.727485e+09            1082.0   
3               1.727485e+09           19702.0   
4                        NaN               NaN   
...                      ...               ...   
1739            1.727570e+09           20500.0   
1740            1.727570e+09            1224.0   
1741            1.727570e+09               NaN   
1742            1.727570e+09               NaN   
1743            1.727571e+09            1664.0   

                    fixture.venue.name             fixture.venue.city  ...  \
0     Estadio Juan Bautista Gargantini  Mendoza, Provincia de Mendoza  ...   
1               National Sports Center              Blaine, Minnesota  ...   
2                   Estadio Cuauhtémoc                         Puebla  ...   
3     Estadio Rogelio Silvino Livieres                       Asunción  ...   
4                                 None                           None  ...   
...                                ...                            ...  ...   
1739                  Inter&Co Stadium               Orlando, Florida  ...   
1740       Estadio Luís Alfonso Giagni                       Asunción  ...   
1741  Rocco B. Commisso Soccer Stadium                  New York City  ...   
1742                    Endeavour Park                        Porirua  ...   
1743      Estadio Rafael Agustín Tovar                        Barinas  ...   

     goals.home goals.away  score.halftime.home  score.halftime.away  \
0           1.0        1.0                  1.0                  1.0   
1           1.0        0.0                  0.0                  0.0   
2           2.0        3.0                  0.0                  1.0   
3           0.0        0.0                  0.0                  0.0   
4           NaN        NaN                  NaN                  NaN   
...         ...        ...                  ...                  ...   
1739        3.0        1.0                  1.0                  0.0   
1740        1.0        1.0                  1.0                  0.0   
1741        1.0        0.0                  0.0                  0.0   
1742        0.0        4.0                

*we have a total of 1,744 soccer matches played on the 28th of sept, 2024*

In [26]:
#Let's view the column structure for our dataframe
fixturesDF.columns

Index(['fixture.id', 'fixture.referee', 'fixture.timezone', 'fixture.date',
       'fixture.timestamp', 'fixture.periods.first', 'fixture.periods.second',
       'fixture.venue.id', 'fixture.venue.name', 'fixture.venue.city',
       'fixture.status.long', 'fixture.status.short', 'fixture.status.elapsed',
       'league.id', 'league.name', 'league.country', 'league.logo',
       'league.flag', 'league.season', 'league.round', 'teams.home.id',
       'teams.home.name', 'teams.home.logo', 'teams.home.winner',
       'teams.away.id', 'teams.away.name', 'teams.away.logo',
       'teams.away.winner', 'goals.home', 'goals.away', 'score.halftime.home',
       'score.halftime.away', 'score.fulltime.home', 'score.fulltime.away',
       'score.extratime.home', 'score.extratime.away', 'score.penalty.home',
       'score.penalty.away'],
      dtype='object')

### Filter for 90-mins matches

Our goal is to find the percetange of teams that scored more or less than one(1) goal in an entire 90-mins football match for the day

First, lets Filter for Full Time fixtures: 
    i.e response.status.short == FT

In [87]:
# Lets filter for FullTime Matches only

ftFixturesDF = fixturesDF[fixturesDF['fixture.status.short'] == 'FT']
ftFixturesDF

fixture.id             fixture.referee fixture.timezone  \
0        1158871                   J. Baliño              UTC   
1        1174518                    J. Matto              UTC   
2        1206084  Oscar Mejia Garcia, Mexico              UTC   
3        1223155                  J. Benítez              UTC   
5        1300427                        None              UTC   
...          ...                         ...              ...   
1739     1168633                 I. Petrunok              UTC   
1740     1223151                        None              UTC   
1741     1285218                        None              UTC   
1742     1306667                        None              UTC   
1743     1234120                        None              UTC   

                   fixture.date  fixture.timestamp  fixture.periods.first  \
0     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
1     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
2     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
3     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
5     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
...                         ...                ...                    ...   
1739  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1740  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1741  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1742  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1743  2024-09-28T23:45:00+00:00         1727567100           1.727567e+09   

      fixture.periods.second  fixture.venue.id  \
0               1.727485e+09              83.0   
1               1.727485e+09           20581.0   
2               1.727485e+09            1082.0   
3               1.727485e+09           19702.0   
5               1.727485e+09           11251.0   
...                      ...               ...   
1739            1.727570e+09           20500.0   
1740            1.727570e+09            1224.0   
1741            1.727570e+09               NaN   
1742            1.727570e+09               NaN   
1743            1.727571e+09            1664.0   

                    fixture.venue.name             fixture.venue.city  ...  \
0     Estadio Juan Bautista Gargantini  Mendoza, Provincia de Mendoza  ...   
1               National Sports Center              Blaine, Minnesota  ...   
2                   Estadio Cuauhtémoc                         Puebla  ...   
3     Estadio Rogelio Silvino Livieres                       Asunción  ...   
5                Estadio Tres de Marzo                        Zapopan  ...   
...                                ...                            ...  ...   
1739                  Inter&Co Stadium               Orlando, Florida  ...   
1740       Estadio Luís Alfonso Giagni                       Asunción  ...   
1741  Rocco B. Commisso Soccer Stadium                  New York City  ...   
1742                    Endeavour Park                        Porirua  ...   
1743      Estadio Rafael Agustín Tovar                        Barinas  ...   

     goals.home goals.away  score.halftime.home  score.halftime.away  \
0           1.0        1.0                  1.0                  1.0   
1           1.0        0.0                  0.0                  0.0   
2           2.0        3.0                  0.0                  1.0   
3           0.0        0.0                  0.0                  0.0   
5           3.0        2.0                  0.0                  1.0   
...         ...        ...                  ...                  ...   
1739        3.0        1.0                  1.0                  0.0   
1740        1.0        1.0                  1.0                  0.0   
1741        1.0        0.0                  0.0                  0.0   
1742        0.0        4.0                

*we have a total of 1,384 90-mins soccer matches played on the 28th of sept, 2024*

In [88]:
#let's highlight our columns of focus

ftFixturesDF[['goals.home', 'goals.away', 'score.halftime.home', 'score.halftime.away', 'score.fulltime.home', 'score.fulltime.away']].head(10)

goals.home  goals.away  score.halftime.home  score.halftime.away  \
0          1.0         1.0                  1.0                  1.0   
1          1.0         0.0                  0.0                  0.0   
2          2.0         3.0                  0.0                  1.0   
3          0.0         0.0                  0.0                  0.0   
5          3.0         2.0                  0.0                  1.0   
7          1.0         1.0                  1.0                  0.0   
9          5.0         0.0                  2.0                  0.0   
10         0.0         0.0                  0.0                  0.0   
11         1.0         0.0                  1.0                  0.0   
12         1.0         1.0                  1.0                  0.0   

    score.fulltime.home  score.fulltime.away  
0                   1.0                  1.0  
1                   1.0                  0.0  
2                   2.0                  3.0  
3                   0.0                  0.0  
5                   3.0                  2.0  
7                   1.0                  1.0  
9                   5.0                  0.0  
10                  0.0                  0.0  
11                  1.0                  0.0  
12                  1.0                  1.0

In [89]:
# Are there cases where fulltime score is not equal goals scored
ftFixturesDF[(ftFixturesDF['goals.home'] != ftFixturesDF['score.fulltime.home']) | 
                 (ftFixturesDF['goals.away'] != ftFixturesDF['score.fulltime.away'])]

fixture.id fixture.referee fixture.timezone               fixture.date  \
1121     1299862            None              UTC  2024-09-28T14:00:00+00:00   

      fixture.timestamp  fixture.periods.first  fixture.periods.second  \
1121         1727532000           1.727532e+09            1.727536e+09   

      fixture.venue.id fixture.venue.name fixture.venue.city  ... goals.home  \
1121            4942.0  Broadwood Stadium        Cumbernauld  ...        0.0   

     goals.away  score.halftime.home  score.halftime.away score.fulltime.home  \
1121        3.0                  NaN                  NaN                 0.0   

     score.fulltime.away score.extratime.home score.extratime.away  \
1121                 2.0                  NaN                  NaN   

      score.penalty.home score.penalty.away  
1121                 NaN                NaN  

[1 rows x 38 columns]

*Found a case but am not so worried about it*

In [90]:
# check for where goals.home + goals.away >= 1

over0_5DF = ftFixturesDF[(ftFixturesDF['goals.home'] + ftFixturesDF['goals.away']) >= 1]
under0_5DF = ftFixturesDF[(ftFixturesDF['goals.home'] + ftFixturesDF['goals.away']) < 1]

In [91]:
#we want to ensure the entire columns are processed

len(ftFixturesDF) - (len(over0_5DF) + len(under0_5DF))

0

In [34]:
len(over0_5DF)

1100

In [36]:
len(under0_5DF)

90

*we can see we have 1,100 over 0.5 and only 90 under 0.5s*

In [93]:
# what is the percentage of over 0.5s compared to under0.5

percentage = (len(over0_5DF) / len(ftFixturesDF)) * 100
print(f"percentage win of over 0.5 on {date} is {percentage:.2f}%")

percentage win of over 0.5 on 2024-09-28 is 92.47%


*percentage win of over 0.5 on 2024-09-28 is 92.47% therefore under 0.5 must be: 7.53%*

### New Goal - Check for over_0.5 after the first half

 *To do this, we can check for over0.5 in second half at 0-0 first half score

In [95]:
#get games with first half draw

firstHalfDrawDF = ftFixturesDF[(ftFixturesDF['score.halftime.home'] + ftFixturesDF['score.halftime.away']) == 0]
firstHalfDrawDF

fixture.id                          fixture.referee fixture.timezone  \
1        1174518                                 J. Matto              UTC   
3        1223155                               J. Benítez              UTC   
10       1183306       Emerson Ricardo de Almeida Andrade              UTC   
13       1306665                                     None              UTC   
15       1206085  Jorge Abraham Camacho Peregrina, Mexico              UTC   
...          ...                                      ...              ...   
1734     1151194                        Ramy Touchan, USA              UTC   
1735     1151195                   Villarrea Armando, USA              UTC   
1736     1151196                  Marcos de Oliveira, USA              UTC   
1738     1151198                       Fotis Bazakos, USA              UTC   
1741     1285218                                     None              UTC   

                   fixture.date  fixture.timestamp  fixture.periods.first  \
1     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
3     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
10    2024-09-28T00:30:00+00:00         1727483400           1.727483e+09   
13    2024-09-28T00:30:00+00:00         1727483400           1.727483e+09   
15    2024-09-28T01:00:00+00:00         1727485200           1.727485e+09   
...                         ...                ...                    ...   
1734  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1735  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1736  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1738  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1741  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   

      fixture.periods.second  fixture.venue.id  \
1               1.727485e+09           20581.0   
3               1.727485e+09           19702.0   
10              1.727487e+09             230.0   
13              1.727487e+09           19314.0   
15              1.727489e+09           20476.0   
...                      ...               ...   
1734            1.727570e+09           20764.0   
1735            1.727570e+09             310.0   
1736            1.727570e+09            1618.0   
1738            1.727570e+09            1620.0   
1741            1.727570e+09               NaN   

                       fixture.venue.name         fixture.venue.city  ...  \
1                  National Sports Center          Blaine, Minnesota  ...   
3        Estadio Rogelio Silvino Livieres                   Asunción  ...   
10    Estádio Major Antônio Couto Pereira           Curitiba, Paraná  ...   
13                            Fraser Park                 Lower Hutt  ...   
15                    Estadio Corregidora      Santiago de Querétaro  ...   
...                                   ...                        ...  ...   
1734                        Chase Stadium   Fort Lauderdale, Florida  ...   
1735                         Stade Saputo           Montreal, Quebec  ...   
1736                     Gillette Stadium  Foxborough, Massachusetts  ...   
1738                          Subaru Park      Chester, Pennsylvania  ...   
1741     Rocco B. Commisso Soccer Stadium              New York City  ...   

     goals.home goals.away  score.halftime.home  score.halftime.away  \
1           1.0        0.0                  0.0                  0.0   
3           0.0        0.0                  0.0                  0.0   
10          0.0        0.0                  0.0                  0.0   
13          0.0        2.0                  0.0                  0.0   
15          0.0        0.0                  0.0                  0.0   
...         ...        ...                  ...                  ...   
1734        1.0        1.0                  0.0                  0.0   
1735        3.0        0.0                

In [68]:
len(firstHalfDrawDF)

400

*we have only 400 games with first half draw*

In [96]:
# Lets find games with over0.5 after halftime

O_5_2HalfDF = firstHalfDrawDF[(firstHalfDrawDF['goals.home'] + firstHalfDrawDF['goals.away']) >= 1]
O_5_2HalfDF

fixture.id          fixture.referee fixture.timezone  \
1        1174518                 J. Matto              UTC   
13       1306665                     None              UTC   
16       1237180                     None              UTC   
17       1245498       Y. Peinado Aguirre              UTC   
28       1299759                     None              UTC   
...          ...                      ...              ...   
1734     1151194        Ramy Touchan, USA              UTC   
1735     1151195   Villarrea Armando, USA              UTC   
1736     1151196  Marcos de Oliveira, USA              UTC   
1738     1151198       Fotis Bazakos, USA              UTC   
1741     1285218                     None              UTC   

                   fixture.date  fixture.timestamp  fixture.periods.first  \
1     2024-09-28T00:00:00+00:00         1727481600           1.727482e+09   
13    2024-09-28T00:30:00+00:00         1727483400           1.727483e+09   
16    2024-09-28T01:00:00+00:00         1727485200           1.727485e+09   
17    2024-09-28T01:00:00+00:00         1727485200           1.727485e+09   
28    2024-09-28T02:30:00+00:00         1727490600           1.727491e+09   
...                         ...                ...                    ...   
1734  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1735  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1736  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1738  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   
1741  2024-09-28T23:30:00+00:00         1727566200           1.727566e+09   

      fixture.periods.second  fixture.venue.id  \
1               1.727485e+09           20581.0   
13              1.727487e+09           19314.0   
16              1.727489e+09             408.0   
17              1.727489e+09            1076.0   
28              1.727494e+09               NaN   
...                      ...               ...   
1734            1.727570e+09           20764.0   
1735            1.727570e+09             310.0   
1736            1.727570e+09            1618.0   
1738            1.727570e+09            1620.0   
1741            1.727570e+09               NaN   

                    fixture.venue.name         fixture.venue.city  ...  \
1               National Sports Center          Blaine, Minnesota  ...   
13                         Fraser Park                 Lower Hutt  ...   
16        Estadio Rafael Ángel Camacho                  Turrialba  ...   
17                       Estadio AKRON                    Zapopan  ...   
28                      Campo Quirazco            Xalapa-Enríquez  ...   
...                                ...                        ...  ...   
1734                     Chase Stadium   Fort Lauderdale, Florida  ...   
1735                      Stade Saputo           Montreal, Quebec  ...   
1736                  Gillette Stadium  Foxborough, Massachusetts  ...   
1738                       Subaru Park      Chester, Pennsylvania  ...   
1741  Rocco B. Commisso Soccer Stadium              New York City  ...   

     goals.home goals.away  score.halftime.home  score.halftime.away  \
1           1.0        0.0                  0.0                  0.0   
13          0.0        2.0                  0.0                  0.0   
16          2.0        0.0                  0.0                  0.0   
17          2.0        2.0                  0.0                  0.0   
28          0.0        1.0                  0.0                  0.0   
...         ...        ...                  ...                  ...   
1734        1.0        1.0                  0.0                  0.0   
1735        3.0        0.0                  0.0                  0.0   
1736        1.0        0.0                  0.0                  0.0   
1738        1.0        1.0                  0.0                  0.0   
1741        1.0        0.0                  0.0       

*we have only 297 of that with over 0.5*

Therefore percentage of over 0.5 after first half is approximately 75% that is about 20% dip compared to fulltime over0.5

This result allows us to conclude that the average chances for an event to occur chages as the events goes inplay.

As such we can use convergent analysis to measure the odds of an event occuring and we can measure these events in every mins of the match.

## Intro to Convergent Analysis

My idea of convergent analysis has not be fully reasearched and developed however I will love to introduce the notion in this notebook.

### Theory: 
In the presence of large data, patterns emerges.

In the case of a soccer match, the idea is to find out the likely of a football event to occur and in particular the total numbers of goals that can be scored in a merge.

We can approach this problem by collecting a very large amount of football match data, lets say over 1million or 2million matches over a long period of time,

This data can be considering including only a couple of competions to keep the focus narrower.

We can randomize the rows of the football matches then split the table into 2s.

we can specify an interval and check 

### Approach to convergent Analysis
1. Have a large dataset
2. Randonmise dataset
2. split data set into n-places where each dataset is still very large
3. Define an interval k
4. For each data-set analyse the first-kth rows
5. Analyse from row 1 - 2k rows then compare the statistics with the previous analysis
6. Continue this step till statistics starts to converge.
7. If the statistics of all the splits converge around the same value then that value is the natural probablity of such event occuring


## Intro to Convergent Analysis

My idea of convergent analysis has not been fully researched and developed; however, I would like to introduce the notion in this notebook.

### Theory: 
In the presence of large data, patterns emerge.

In the context of a soccer match, the goal is to determine the likelihood of a football event occurring, particularly the total number of goals that may be scored in a match.

We can approach this problem by collecting a significant amount of football match data—let's say over 1 or 2 million matches spanning a long period. This data would ideally focus on only a few specific competitions, keeping the scope narrow.

Once the data is collected, we can randomize the rows of the football matches and split the table into two or more subsets.

### Approach to Convergent Analysis:
1. **Obtain a large dataset**: The dataset should include match statistics and relevant features such as teams, goals, date of match, and competition type.
2. **Randomize the dataset**: Shuffle the rows to eliminate any ordering bias.
3. **Split the dataset into `n` parts**: Ensure that each subset still contains a significant number of matches.
4. **Define an interval `k`**: This interval represents a batch of rows (matches) for analysis.
5. **Analyze the first `k` rows** of each dataset subset: Compute the relevant statistics such as the total number of goals, average goals, or frequency of goals exceeding a threshold (e.g., 0.5 goals).
6. **Analyze rows from `1` to `2k`**: Extend the dataset analyzed and compare the statistics with the previous batch (first `k` rows).
7. **Continue the process**: Keep increasing the interval, comparing the new statistics with the previous results, until the statistics of all splits start to converge.

When the statistics stabilize or converge across the different subsets of data, this value can be considered the **natural probability** of the event occurring.

### Example in Football Match Analysis:
For instance, if we are trying to determine the likelihood of football matches having over 0.5 goals, we can use convergent analysis to find the natural probability of this event. After analyzing the statistics of subsets of matches, if the probabilities of goals totaling over 0.5 converge around a certain value, we can infer that this is the natural probability of such an event occurring. This could be further extended to other goal thresholds, such as over 1.5 goals, over 2.5 goals, and so on.